### PR TITLE
 add krel sign blobs and images commands

### LINKED ISSUE #2618: Implement file signing step
## Background

As part of the [Signing Release Artifacts KEP](https://github.com/kubernetes/enhancements/issues/3031), we need to build file signing into our release pipeline. Up to now, we had considered building signing capabilities directly into krel. However, thinking a little bit more into the architecture of the pipeline it is possible we may end up not needing to work on it and instead could use standard cosign to sign the files.

Up to now, we had not considered that we need to sign our artifacts outside of `krel stage`. We do not only need to build a new step in the pipeline to sign files after they are built and staged, but we also need to move the current signing image signing logic out of krel (see https://github.com/kubernetes/release/issues/2611 for a more broad explanation of why).

## Our choice: `krel sign` or vanilla `cosign`

This means we have two choices to implement the file signing step:

1. [Build file signing into krel](https://github.com/kubernetes/release/issues/2600), eg a `krel sign` subcommand
2. Use a standard [`cosign` release](https://github.com/sigstore/cosign/releases) to sign the files

We should only work on # 1 if there is a compelling reason to do it and, frankly, I'm failing to see one.

## Architecture

In order to better understand how the signing step fits into the release process, let's take a look at how we sign images today:

```mermaid
flowchart TB
  subgraph krel["<h2>krel (today)</h2>"]
    direction TB
    style krel fill:#eee,stroke:#ccc,stroke-width:2px,font-weight:bold,font-size:120%

    subgraph staging["<h3>staging</h3>"]
      s1[build]-->s2["sign container images"]
    end
    subgraph promotion["<h3>image promotion</h3>"]
      ip1["k8s-org signature"]-->ip2
      ip2["signature replication"]
    end
    subgraph release["<h3>release</h3>"]
      r1["check image signatures"]-->r2["release"]
    end
    staging-->promotion
    promotion-->release
end
````

The full proposal to sign files (and ultimately images too) would look like the following. This introduces a new signing step to handle the internal RelEng signatures. Initially would handle files but we should move the signing logic from krel to this new step too to restrict access to the signing account:

```mermaid
flowchart TB
  subgraph krel["<h2>krel (new)</h2>"]
    direction TB
    style krel fill:#eee,stroke:#ccc,stroke-width:2px,font-weight:bold,font-size:120%
    style fpromotion fill:#eee,stroke:#ccc,color:ccc,stroke-width:2px,font-weight:bold,font-size:120%
    style a fill:#B3D5FF,stroke:#ccc,color:003C84,stroke-width:2px,font-weight:bold,font-size:120%
    style fp1 fill:#eee,stroke:#ccc,color:ccc,stroke-width:2px,font-weight:bold,font-size:120%

    subgraph staging["<h3>staging</h3>"]
      s1[build]-->s2["stage images and files"]
    end
    subgraph signing["<h3>signing</h3>"]
      sign1["read staging SBOM<br>to get artifacts that<br>need to be signed"]
      sign1-->sign2["sign files"]
      sign1-->sign3["sign images"]
    end
    subgraph ipromotion["<h3>image promotion</h3>"]
      ip1["k8s-org signature"]-->ip2
      ip2["signature replication"]
    end
    subgraph fpromotion["<h3>file promotion</h3><br>(in an indeterminate future)"]
      a["File promotion is out of scope<br>for now but this is<br>where we should fit it"]
      fp1["k8s-org signature on files"]
    end
    subgraph release["<h3>release</h3>"]
      r1["check image signatures"]
      r2["check file signatures"]
      r1 & r2 --> r3["release"]
    end
    staging-->signing
    signing-->ipromotion
    signing-->fpromotion
    ipromotion & fpromotion -->release
end

````

## Should we cancel the development of our file signing codebase?

Short answer: No.

By no means this plan means we need to stop working on our file signing libraries. In the near future, [kpromo](https://github.com/kubernetes-sigs/promo-tools) should sign files just as it signs container images today. File signing in the release process should be seen as the Release Engineering stamp of origin of our artifacts and we should strive to get files stamped with a Kubernetes-wide signature.

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- add krel sign command to sign artifacts

/assign @saschagrunert @puerco @xmudrii @jeremyrickard

#### Which issue(s) this PR fixes:

Fixes #2618 

#### Does this PR introduce a user-facing change?

```release-note
add krel sign blobs and images commands
```

### CODE REVIEW

**Comment by cpanato** [comment]
opening as draft to collect early feedback

**Review by saschagrunert — COMMENTED** [review]
Great start!

File: cmd/krel/cmd/sign.go
Comment (cpanato) [reply cmd/krel/cmd/sign.go:63]: sounds great, will refactor

File: cmd/krel/cmd/sign.go
Comment (cpanato) [reply cmd/krel/cmd/sign.go:120]: roger that

File: cmd/krel/cmd/sign.go
Comment (cpanato) [reply cmd/krel/cmd/sign.go:63]: @saschagrunert should be `files` or maybe `artifacts`?

File: cmd/krel/cmd/sign.go
Comment (cpanato) [reply cmd/krel/cmd/sign.go:143]: as always your reviews are awesome, thanks!

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [reply cmd/krel/cmd/sign.go:63]: I'd assume that artifacts can be files or containers. We could also think about `blobs` but this sounds even more generic than `files`.

Comment (cpanato) [reply cmd/krel/cmd/sign.go:63]: agree, blobs are better
thanks

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign.go:111]: Do we wanna move those to `sign blobs`?

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign.go:37]: We probably need different options for blobs and images.

File: cmd/krel/cmd/sign_blobs.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign_blobs.go:28]:
```suggestion
	"sigs.k8s.io/release-sdk/sign"
```

File: cmd/krel/cmd/sign_blobs.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign_blobs.go:69]: Let's create the signer outside of the goroutine in Line 55 to be able to reuse it.

File: cmd/krel/cmd/sign_blobs.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign_blobs.go:99]: Do we need to check that? I assume the sign API will error if the file does not exist.

**Comment by saschagrunert** [comment]
> > Is krel sign something release managers have to execute manually or do we want to rely on some automation?
> 
> I think the idea is to use it as part of the pipeline for the release, maybe @puerco can comment as well, but this was my understanding

So I assume it has to be run manually in the same way we do it with `krel stage/release`.

> Also, I am having trouble using the lib to use in keyless mode, there is anything we need to do? or does that only work if we set up a GCP service account?

Hm, yeah this could be tested via cloudbuild.yaml.

> also:
> 
> * having some trouble to understand what kind of cloudbuild job we need

Assuming that we run it side by side to krel stage/release, then we have to create a new cloudbuild job in https://github.com/kubernetes/release/tree/master/gcb 

> * and the boilerplate in anago as well

That's tricky, but we can iterate on it. Krel decides based on the flags if it should submit a cloudbuild job or run the actual logic:

https://github.com/kubernetes/release/blob/ae9fe36e23ad25e37101a765cd37127938f54c87/cmd/krel/cmd/stage.go#L136-L143

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:69]: i tried but then the output signature and certificate got the same name :/

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:99]: maybe not, just to fail fast in this case

**Comment by puerco** [comment]
@saschagrunert :
> So I assume it has to be run manually in the same way we do it with krel stage/release.

No I think we don't need to have it triggered by a release manager, I agree with @cpanato that we should add it as part of the existing pipeline:

>  the idea is to use it as part of the pipeline for the release

My thinking is that we should sign everything in a step _inside_ the release GCB job. Specifically, I think we should insert it here:

https://github.com/kubernetes/release/blob/master/gcb/release/cloudbuild.yaml#L44

**Comment by saschagrunert** [comment]
> My thinking is that we should sign everything in a step _inside_ the release GCB job. Specifically, I think we should insert it here:
> 
> https://github.com/kubernetes/release/blob/master/gcb/release/cloudbuild.yaml#L44

Sounds good, let's do it that way to keep things simple.

File: cmd/krel/cmd/sign_blobs.go
Comment (saschagrunert) [reply cmd/krel/cmd/sign_blobs.go:69]: Have you tried to only set `signerOpts.OutputCertificatePath` and `signerOpts.OutputSignaturePath` within the loop?

Comment (saschagrunert) [reply cmd/krel/cmd/sign_blobs.go:69]: Got it, it won't work.

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign.go:31]: Let's just name that `output-path`

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign.go:30]: We do not need that flag if we just use `output-path`.

**Comment by cpanato** [comment]
Run this command in a duplicated kubernetes bucket and all files get signed and both signatures/certificates were pushed to the bucket as well

after the initial review i will finish this pr adding the cloudbuild steps

File: cmd/krel/cmd/sign_blobs.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign_blobs.go:172]: For what do we need this?

**Comment by saschagrunert** [comment]
@cpanato do you need support with that, I'd like to get it merged sooner than later to give it a test before 1.26 got released.

**Comment by cpanato** [comment]
> @cpanato do you need support with this one? I'd like to get it merged sooner than later to give it a test before 1.26 got released.

will continue on this today, just planning my day

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:172]: maybe not, just to slow down a bit

File: cmd/krel/cmd/sign_blobs.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign_blobs.go:203]: This will panic if there are no args provided, we have to check the `len(args)` as well.

File: cmd/krel/cmd/sign.go
Comment (cpanato) [reply cmd/krel/cmd/sign.go:52]: great catch! thanks

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:39]: copy paste issues :)

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:110]: no, again copy and paste issues, it is clear i dont know how to copy and reformat :/

File: cmd/krel/cmd/sign_blobs.go
Comment (xmudrii) [inline cmd/krel/cmd/sign_blobs.go:188]: I have some concerns that this might override binaries in some cases. For example, what if you have two binaries named `kubectl` and `kubectl-convert`? In that case, you're going to copy:
- `kubectl*`
- `kubectl-convert*`

The first one might copy:
- `kubectl.sig`
- `kubectl.cert`
- `kubectl-convert` <-- this is another binary itself
- `kubectl-convert.sig` <-- this is a signature of another binary
- `kubectl-convert.cert` <-- this is a certificate of another binary

This should be a no-op, but I'm wondering if we should do this at all or if this can cause any failure.

File: gcb/release/cloudbuild.yaml
Comment (cpanato) [reply gcb/release/cloudbuild.yaml:56]: dont needed

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:188]: great catch, lets filter with .sig and .cert?

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:123]: fixed, i dont know why i was using containsany

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:202]: thanks for your review @xmudrii those are very valuable <3

File: cmd/krel/cmd/sign_blobs.go
Comment (xmudrii) [reply cmd/krel/cmd/sign_blobs.go:188]: > lets filter with .sig and .cert?

Yes, that makes sense to me.

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go:202]: done

File: cmd/krel/cmd/sign_blobs.go
Comment (cpanato) [reply cmd/krel/cmd/sign_blobs.go]: done

**Review by xmudrii — APPROVED** [review]
Great job, @cpanato!!
/lgtm
/approve
/hold
if @saschagrunert wants to take a look, otherwise unhold

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign.go:41]:
```suggestion
// signCmd represents the subcommand for `krel sign`
```

File: cmd/krel/cmd/sign.go
Comment (saschagrunert) [inline cmd/krel/cmd/sign.go:44]: nit:
```suggestion
	Short:         "sign images and blobs",
```

**Review by saschagrunert — APPROVED** [review]
Just two nits, otherwise LGTM

File: cmd/krel/cmd/sign.go
Comment (cpanato) [reply cmd/krel/cmd/sign.go:44]: thanks
PTAL

**Review by saschagrunert — APPROVED** [review]
/unhold

Thank you @cpanato!

**Comment by saschagrunert** [comment]
@cpanato do we wanna give this a test run ahead of the next RC?

**Comment by saschagrunert** [comment]
@cpanato sure

#### Stage

```
> go run ./cmd/krel stage --type rc --branch=release-1.26
```
https://console.cloud.google.com/cloud-build/builds/3016c99d-2552-47c9-b117-f79cd8927cab?project=648026197307

**Pull request merged**
